_posts/2021-05-02-elasticsearch-basics.md

```diff
@@ -0,0 +1,131 @@
+---
+title: 'Introduction to Elasticsearch Architecture'
+date: 2021-05-02 07:30:00
+---
+
+Elasticsearch is a distributed storage engine optimized for search operations.
+
+Data is stored in Elasticsearch as JSON documents organized in indices. In older versions (prior 7.x) indices was additionaly splitted into types, but they were dropped.
+
+Indices can define mappings (like table schema in SQL) for stored documents, but it is not mandatory. It is possible to set mapping validation as strict - meaning only documents matching the mapping can be indexed (saved) into the database.
+
+One Elasticsearch instance (cluster) can contain a number of indices. One index is built on top of multiple shards. Shard is a minimal non-dividable piece of storage. Shards of one index can be physically located on different machines (nodes).
+
+When defining and index one must define how many shards and replicas it should be built of. While it's possible to change replica factor on the ongoing index, it is not possible to change the number of primary shards, so you must think carefully what would be the best number for you.
+
+Elasticsearch should be started on multiple nodes, preferably at least 3. That number would make it easy to ensure high availability on the network errors.
```

Comment:
maybe to be clear or underline it, start with sth like this:
"(It is recommended that) production environment should consist of at least 3 nodes."